src/Action/SetObjectFieldValueAction.php

```diff
@@ -115,7 +115,18 @@ public function __invoke(Request $request): JsonResponse
 
         // Handle date and datetime types have setter expecting a DateTime object
         if ('' !== $value && \in_array($fieldDescription->getType(), ['date', 'datetime'], true)) {
-            $value = new \DateTime($value);
+            $inputTimezone = new \DateTimeZone(date_default_timezone_get());
+            $outputTimezone = $fieldDescription->getOption('timezone') ?: date_default_timezone_get();
+
+            if (!$outputTimezone instanceof \DateTimeZone) {
+                $outputTimezone = new \DateTimeZone($outputTimezone);
+            }
+
+            $value = new \DateTime($value, $outputTimezone);
+
+            if ($outputTimezone->getName() !== $inputTimezone->getName()) {
+                $value->setTimezone($inputTimezone);
+            }
```

Comment:
```suggestion
            if ($timezone = $fieldDescription->getOption('timezone')) {
                $inputTimezone = $timezone instanceof \DateTimeZone ? $timezone : new \DateTimeZone($timezone);
                $value->setTimezone($inputTimezone);
            }
```

Reply:
This is the wrong code.

Reply:
Maybe I'm missing something. My suggestion is based on this fact: https://3v4l.org/UF1Rp.
How calling `date_default_timezone_get()` is different?

Reply:
Your code does not do what is needed here.

Reply:
Please review again.

Reply:
Besides. Assigning a value to a variable in a condition is a sign of a bad coding style. It is worth avoiding this.

Reply:
There is no need to perform the assign operation inside the `if` condition :+1:. Nevertheless, that isn't changing the result.
I'll be back with some feedback when I'm able to check it locally.

Reply:
I've executed the test `SetObjectFieldValueActionTest` locally, and I can't find any difference in the results using the original code or your changes.
I'd suggest to create a representative test exposing what this PR fixes.

Reply:
I have added tests for this change.